src/NuGet.Core/NuGet.Packaging/Signing/Archive/EndOfCentralDirectoryRecord.cs

```diff
@@ -47,5 +46,65 @@ internal static EndOfCentralDirectoryRecord Read(BinaryReader reader)
 
             return header;
         }
+
+        private static void SeekToEndOfCentralDirectoryRecord(BinaryReader reader)
+        {
+            var byteSignature = BitConverter.GetBytes(Signature);
+            var length = reader.BaseStream.Length;
+
+            if (length < byteSignature.Length)
+            {
+                ThrowByteSignatureNotFoundException(byteSignature);
+            }
+
+            const int DefaultBufferSize = 4096;
+
+            var bufferSize = (int)Math.Min(length, DefaultBufferSize);
+            var matchingByteCount = 0;
+
+            // position is 0-based, while bufferSize is 1-based.
+            for (var position = length - bufferSize; position >= 0; position -= bufferSize)
+            {
+                reader.BaseStream.Position = position;
+
+                var buffer = reader.ReadBytes(bufferSize);
+
+                for (var i = buffer.Length - 1; i >= 0; --i)
+                {
+                    if (buffer[i] == byteSignature[byteSignature.Length - matchingByteCount - 1])
+                    {
+                        ++matchingByteCount;
+
+                        if (matchingByteCount == byteSignature.Length)
```

Comment:
could there be a case where there could be potentially more matchingBytes but we are exiting too early?
probably not a valid scenario but still curious.

Reply:
That would be a bug.  I don't think there's a risk of this though.  If `matchingByteCount == byteSignature.Length` it means we found the byte signature in a contiguous block.  Any time a non-matching byte is found, `matchingByteCount` is reset to 0.

We're reading backwards from the end of stream in adjacent, non-overlapping chunks of size `bufferSize` and searching backwards for the EOCD byte signature.  In the 99th percentile we'll find the EOCD byte signature in the first 4,096 buffer read.
```
 50 4b 05 06
|---buffer---|
```
However, in some unusual cases, the byte signature may be in the non-last buffer:
```
 50 4b 05 06
|---buffer---|---buffer---|
```
...or may span buffers.
```
        50 4b 05 06
|---buffer---|---buffer---|
```

Once we find the EOCD byte signature, we're done.  How can there be more matching bytes after this?

You might be thinking:  what if there are multiple EOCD byte signatures.  Aside from garbage data or malformed ZIP, there's a few ways that could theoretically happen.  In addition to the EOCD byte signature:

1.  The ZIP file comment also contains the EOCD byte sequence.
1.  Some data after the ZIP contains the EOCD byte sequence.

Either case could be intentional (e.g.:  specifically putting the EOCD sequence in the ZIP file comment or postpending one ZIP to another) or unintentional unluckiness that whatever comment/extra data you've postpended to your ZIP happens to contain the same 4-byte sequence as the EOCD byte signature.  In the former case, we'll see the first EOCD from the end as "the EOCD," which could be wrong, and we'll fail reading the rest of the ZIP.  In the latter case, you're supremely (un-)lucky (1 in 4 billion) to have encountered such a phenomenon and should have bought a lottery ticket instead.

The previous code had the same drawbacks with respect to multiple EOCD byte signatures.

Reply:
thanks for the detailed explanation!